# Convolutional Neural Networks with keras


This will illustrate the basics of a convolutional neural networks using keras to create a digit classification network. WITHOUT A LOT OF MATH

## What is a Convolutional whatnot?

A better question is this: 

Is the image below a dog or a cat?

![Dog Image](https://www.what-dog.net/Images/faces2/scroll0015.jpg)


Obviously it's a dog!



It took you, a human, no time at all to say that that is a dog. Our brains are cool like that. So how did you know that's a dog? Well, you were taught what dogs look like, vs what cats look like, by your parents. When you asked what that fluffy thing was, they told you it was a dog, and the same goes for the cat.

Ok, easy, teach babies what cats and dogs are. Not that hard, right? Well, yes. Our brains, even at a young age are var more advanced than computers. So how would you teach a computer?


## Computers vs. Humans: Vision

If you know the physics behind our eyes, you will know that our eyes take in wavelengths of light that bounce off objects around us, and our brains interpret it. It's not hard for us to see because that process is done subconsciously. It's very easy to recognize images just by looking at them, but computers don't have the same ability.

A picture like this:
![Lion Image](https://a-z-animals.com/media/animals/images/original/lion7.jpg)

Is easy to recognize as a lion.


For computers it's a bit different. Images are stored as a bunch of numbers.

![Matrix](http://chem4823.usask.ca/images/matrices1.gif)

Try classifiying this, you can't. A, it means nothing, its just a bunch of random numbers I found on Google. B, it's not in a format you understand.

Computers have to take images as a bunch of numbers. Which they can understand. So how can computers classify images? Well they use something called Neural Networks.

## Computers vs. Humans: Networks

The neural network actually stems from biology, and is the basis for how our brain learns.

![Neural Network](https://upload.wikimedia.org/wikipedia/commons/thumb/b/b5/Neuron.svg/1200px-Neuron.svg.png)

The neural network a computer uses follows a similar pattern. Input comes from a previous neuron and gets transferred through, with a certain strength, or the stregth of the synapse between the two neurons.


## Back to Convolutional whatnot...

For classifiy images using a computer, we use a type of neural network called a Convolutional Neural Network (CNN). A CNN iterates over an image and sends pixel values to the network for classification.

![Neural Network Image](http://cs231n.github.io/assets/nn1/neural_net2.jpeg)

![Network Image 2](http://cs231n.github.io/assets/cnn/convnet.jpeg)

## Let's build a CNN with python and keras

In [1]:
# dependencies
from __future__ import print_function

# We're imported our Keras submodules, such as Dense, Dropout, and MaxPooling2D. All things
# we will need to make our network
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten, MaxPooling2D, Conv2D
from keras.datasets import cifar10
from keras.optimizers import rmsprop
from keras.preprocessing.image import ImageDataGenerator
import keras.utils as np_utils
import numpy as np

import os

Using TensorFlow backend.


Now that we've imported our dependencies, we can get started. We're going so define some constants that we'll use later on in our program.

In [2]:
batch_size = 32
number_of_classes = 10
epochs = 200
data_augmentation = True
number_of_predictions = 20
save_directory = os.path.join(os.getcwd(), 'saved_models')
model_name = 'keras_first_cnn_test.h5'

The number_of_classes is the amount of classification classes we want. i.e. 0,1,2,3...10. Each one of those numbers is a class. We also define our save_directory and the model_name that we'll call it.

In [3]:
# We're going to shuffle our test data that keras provides into testing and training data randomly
# The data, shuffled and split between train and test sets:
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# Convert class vectors to binary class matrices.
y_train = np_utils.to_categorical(y_train, number_of_classes)
y_test = np_utils.to_categorical(y_test, number_of_classes)

x_train shape: (50000, 32, 32, 3)
50000 train samples
10000 test samples


### Model Time

We're going to create a Keras sequential model. With this we can add layers easily with the .add() method.

In [4]:
model = Sequential()
model.add(Conv2D(32, (3,3), padding='same', input_shape=x_train.shape[1:]))

Our first layer is out initial convolutional layer. We're going to be following this basic format:

![Hello](https://fr.mathworks.com/content/mathworks/fr/fr/discovery/convolutional-neural-network/jcr:content/mainParsys/image_copy.adapt.full.high.jpg/1506591540823.jpg)


The only exception is that we will be implimenting dropout functions as well as some other additions.


In [5]:
# add an activation layer with ReLU
model.add(Activation('relu'))
# add Dnother convolutional layer
model.add(Conv2D(32, (3,3)))
# add another activation layer now
model.add(Activation('relu'))
# perform maxPooling
model.add(MaxPooling2D(pool_size=(2,2)))
# finally add the dropout layer

What did we just do?

Well we added a ReLU activation layer, another convolutional layer with a filter of 32, and a kernal_size of [3,3], then another ReLU activation, and finally we added a MaxPooling layer with a pool size of [2,2]. Easy... 😉

## Activation Functions

It sounds like something you could say to your teacher to make them think you're some kind of genius, but it's actually quite simple. An activation function mimics the human brain. Inside the neuron, a chemical reaction takes place to determine if the signal should pass to the next neuron. That's what an activation function does in basic.

We're using a ReLU (Rectified Linear Unit) activation function as it's what most closly mimics the brain.

Some other activation functions include:

* Sigmoid: https://en.wikipedia.org/wiki/Sigmoid_function

* Softplus

* Gaussian

* Softmax

Some more info is listed here: https://en.wikipedia.org/wiki/Rectifier_(neural_networks)


## Pooling

So what's pooling?

Pooling is a way to reduce noise in an image in order to make analysis easier for the next layer.

Our pooling layer, MaxPooling2D, using the Max Pooling algorithm to reduce extra information.
![Max Pooling](https://upload.wikimedia.org/wikipedia/commons/e/e9/Max_pooling.png)


It takes the largest value in the matrix and that is the new pixel value of the output matrix.



In [6]:
# add the dropout layer now, and drop out 25%
model.add(Dropout(0.25))

## Dropout Function

The dropout function is interesting because it prevents "overfitting" by doing something that may seem counter intuitive. The dropout function "destroys" a given percentage of the network setting weights back to 0. It's basically like smashing this delicatly constructed piece of art with a hammer. The reasoning for this is that it keeps the model from being to biased towards one type of class. By implementing a dropout, the model can better classify images of dogs, let's say, that it hasn't seen before. Overfitting can reduce the amount of correct classifications on data the network hasn't seen yet. Another analogy is taking the blinders off of a racehorse, now it can see so many more places to run, besides just straight.


In [7]:
# same as before just increase kernal size by 2
model.add(Conv2D(64, (3,3), padding='same'))
model.add(Activation('relu'))
model.add(Conv2D(64, (3,3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(0.25))

In [8]:
## our final layer bunch
model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(number_of_classes))
model.add(Activation('softmax'))


## Flattening a Layer

Here, we flatten our layer and turn the vetor output of our last calculation into a scalar value, for use in the next layer.

## Dense Layers

A dense layer is just another name for a fully connected layer. A dense layer is where all input nodes are connected to all hidden nodes, which are connected to all output nodes. In our Dense layer, we specify our units, which are the dimensions of the output. i.e. (*, 512).

The keras documentation shows

```python
model = Sequential()
model.add(Dense(32, input_shape=(16,)))
# now the model will take as input arrays of shape (*, 16)
# and output arrays of shape (*, 32)


```

## Softmax classification

Our final activation is softmax. Softmax is multinomial logistic regression. Basically, we're trying to calculate the probabilities to which class it is. So for example it might output:

```
0: 0.98
8: 0.70
9: 0.56
...
```

Meaning that the network thinks that with 98% accuracy, the input image is a 0

In [9]:
# initiate RMSProp optimizer
opt = rmsprop(lr=0.0001, decay=1e-6)

## RMS Optimization

According to cs.toronto.edu:

rmsprop: Divide the learning rate for a weight by a running average of the magnitudes of recent gradients for that weight.

We want to minimize the output of our loss function so we're using RMS prop optimization, which is a type of gradient descent

In [10]:
#train the model now using RMSprop
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])

## Cross Entropy

Here, we're using a loss function called categorical cross entropy. We also use the optimizer as the optimizer we previously created `opt`. Our desired metrics are accuracy

More on cross entropy here: https://en.wikipedia.org/wiki/Cross_entropy

In [11]:
# Finally we're going to convert our training data to a float32
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255


In [12]:
# If we're using data augmentation, we are going to fit our model with our training data, and our batch size, as well as our epochs.
# We're also using some validation_data to make sure our model works fine

if not data_augmentation:
    print('Not using data augmentation.')
    model.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=epochs,
              validation_data=(x_test, y_test),
              shuffle=True)
else:
     print('Using real-time data augmentation.')
     datagen = ImageDataGenerator(
        featurewise_center=False,
        samplewise_center=False,
        featurewise_std_normalization=False,
        samplewise_std_normalization=False,
        zca_whitening=False,
        rotation_range=0,
        width_shift_range=0.1,
        height_shift_range=0.1,
        horizontal_flip=True,
        vertical_flip=False
     )
     datagen.fit(x_train)

     # Fit the model on the batches generated by datagen.flow().
     model.fit_generator(datagen.flow(x_train, y_train,
                                     batch_size=batch_size),
                        steps_per_epoch=x_train.shape[0] // batch_size,
                        epochs=epochs,
                        validation_data=(x_test, y_test),
                        workers=4)

Using real-time data augmentation.
Epoch 1/200
1562/1562 [==============================] - 192s - loss: 1.8685 - acc: 0.3140 - val_loss: 1.5456 - val_acc: 0.4410
Epoch 2/200
1562/1562 [==============================] - 191s - loss: 1.5719 - acc: 0.4279 - val_loss: 1.3786 - val_acc: 0.5084
Epoch 3/200
1562/1562 [==============================] - 190s - loss: 1.4486 - acc: 0.4786 - val_loss: 1.2587 - val_acc: 0.5479
Epoch 4/200
1562/1562 [==============================] - 177s - loss: 1.3547 - acc: 0.5154 - val_loss: 1.1963 - val_acc: 0.5741
Epoch 5/200
1562/1562 [==============================] - 176s - loss: 1.2776 - acc: 0.5455 - val_loss: 1.1126 - val_acc: 0.6121
Epoch 6/200
1562/1562 [==============================] - 179s - loss: 1.2126 - acc: 0.5701 - val_loss: 1.0467 - val_acc: 0.6344
Epoch 7/200
1562/1562 [==============================] - 194s - loss: 1.1669 - acc: 0.5861 - val_loss: 1.0360 - val_acc: 0.6345
Epoch 8/200
1562/1562 [==============================] - 196s - loss:

1562/1562 [==============================] - 194s - loss: 0.8458 - acc: 0.7263 - val_loss: 0.7943 - val_acc: 0.7500
Epoch 128/200
1562/1562 [==============================] - 193s - loss: 0.8528 - acc: 0.7253 - val_loss: 0.7413 - val_acc: 0.7670
Epoch 129/200
1562/1562 [==============================] - 198s - loss: 0.8448 - acc: 0.7263 - val_loss: 0.8722 - val_acc: 0.7321
Epoch 130/200
1562/1562 [==============================] - 197s - loss: 0.8581 - acc: 0.7197 - val_loss: 0.7748 - val_acc: 0.7540
Epoch 131/200
1562/1562 [==============================] - 200s - loss: 0.8578 - acc: 0.7232 - val_loss: 0.6820 - val_acc: 0.7845
Epoch 132/200
1562/1562 [==============================] - 202s - loss: 0.8689 - acc: 0.7215 - val_loss: 0.7012 - val_acc: 0.7727
Epoch 133/200
1562/1562 [==============================] - 203s - loss: 0.8726 - acc: 0.7196 - val_loss: 0.7177 - val_acc: 0.7649
Epoch 134/200
1562/1562 [==============================] - 201s - loss: 0.8729 - acc: 0.7198 - val_loss:

NOTE: Executing the above will train the model and will take some time. A few hours with a decent laptop via Jupyter

In [15]:
# Save model and weights
if not os.path.isdir(save_directory):
    os.makedirs(save_directory)
model_path = os.path.join(save_directory, model_name)
model.save(model_path)
print('Saved trained model at %s ' % model_path)

# Score trained model.
scores = model.evaluate(x_test, y_test, verbose=1)
print('Test loss:', scores[0])
print('Test accuracy:', scores[1])

Saved trained model at /Users/ryan/GitHub/AITesting/Misc./saved_models/keras_first_cnn_test.h5 
 9952/10000 [============================>.] - ETA: 0sTest loss: 1.14125301056
Test accuracy: 0.6293


## Accuracy

Our outputed accuracy is about 0.63. Which is terrible, but with better models, we can improve that accuracy.

Original Source code located:
https://github.com/fchollet/keras/blob/master/examples/cifar10_cnn.py